src/Common/Core/Impl/Services/ServiceManager.cs

```diff
@@ -40,24 +33,36 @@ public class ServiceManager : IServiceManager {
             type = type ?? typeof(T);
             Check.ArgumentNull(nameof(service), service);
             Check.InvalidOperation(() => _s.TryAdd(type, service), "Service already exists");
-            ServiceAdded?.Invoke(this, new ServiceContainerEventArgs(type));
+
+            var mainThread = service as IMainThread;
+            if (mainThread != null) {
+                _mainThread = _mainThread ?? mainThread;
+            }
+
             return this;
         }
 
         /// <summary>
         /// Adds on-demand created service
         /// </summary>
-        /// <param name="factory">Optional creator function. If not provided, reflection with default constructor will be used.</param>
-        public virtual IServiceManager AddService<T>(Func<T> factory) where T: class {
+        /// <param name="factory">Service factory</param>
+        public virtual IServiceManager AddService<T>(Func<IServiceManager, T> factory) where T : class {
             _disposeToken.ThrowIfDisposed();
 
-            var type = typeof(T);
-            var lazy = factory != null 
-                ? new Lazy<object>(() => factory()) 
-                : new Lazy<object>(() => Activator.CreateInstance(type));
+            var lazy = new Lazy<object>(() => factory(this));
+            Check.InvalidOperation(() => _s.TryAdd(typeof(T), lazy), "Service already exists");
+            return this;
+        }
+
+        /// <summary>
+        /// Adds on-demand created service
+        /// </summary>
+        /// <param name="type">Type to instantiate with reflection.</param>
+        public virtual IServiceManager AddService(Type type) {
```

Comment:
Actually it is needed for test container that adds mocks for VS global services and needs both instance and different type